### PR TITLE
Repl/multi timezone support

### PR DESCRIPTION
**Pull request by Andy65**

### CODE REVIEW

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/OpenShiftController.cs [hunk 1]
```diff
@@ -527,7 +529,7 @@ internal async Task ProcessOpenShiftsAsync(string isRequestFromLogicApp)
                 KronosSlots = Constants.KronosOpenShiftsSlotCount,
                 SchedulingGroupId = responseModel.SchedulingGroupId,
                 OrgJobPath = orgJobPath,
-                OpenShiftStartDate = this.utility.UTCToKronosTimeZone(responseModel.SharedOpenShift.StartDateTime),
+                OpenShiftStartDate = this.utility.UTCToKronosTimeZone(responseModel.SharedOpenShift.StartDateTime, timeZone),
```
Comment (ajs117) [inline Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/OpenShiftController.cs]: the variable would be better called kronosTimeZone for consistency.

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/OpenShiftRequestController.cs [hunk 2]
```diff
@@ -137,7 +137,10 @@ public class OpenShiftRequestController : Controller
 
             // Step 1c.
             var allRequiredConfigurations = await this.utility.GetAllConfigurationsAsync().ConfigureAwait(false);
-            var kronosTimeZoneId = this.appSettings.KronosTimeZone;
+            var teamMappings = await this.teamDepartmentMappingProvider.GetMappedTeamDetailsAsync(teamsId).ConfigureAwait(false);
+            var teamMapping = teamMappings.FirstOrDefault();
+
+            var kronosTimeZoneId = string.IsNullOrEmpty(teamMapping.KronosTimeZone) ? this.appSettings.KronosTimeZone : teamMapping.KronosTimeZone;
```
Comment (ajs117) [inline Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/OpenShiftRequestController.cs]: should this be string.IsNullOrEmpty(teamMapping?.KronosTimeZone) with the ?

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/SwapShiftController.cs [hunk 3]
```diff
@@ -191,14 +199,10 @@ public static string GenerateNewGuid()
...
                             // Convert shift time to Kronos time zone.
-                            Emp1FromDateTime = this.utility.UTCToKronosTimeZone(
-                                senderShiftDetails.SharedShift.StartDateTime),
-                            Emp1ToDateTime = this.utility.UTCToKronosTimeZone(
-                                senderShiftDetails.SharedShift.EndDateTime),
-                            Emp2FromDateTime = this.utility.UTCToKronosTimeZone(
-                                recipientShiftDetails.SharedShift.StartDateTime),
-                            Emp2ToDateTime = this.utility.UTCToKronosTimeZone(
-                                recipientShiftDetails.SharedShift.EndDateTime),
+                            Emp1FromDateTime = this.utility.UTCToKronosTimeZone(senderShiftDetails.SharedShift.StartDateTime, mappedTeam.KronosTimeZone),
```
Comment (ajs117) [inline Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/SwapShiftController.cs]: do you need to check if mappedTeam is null first?

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs [hunk 4]
```diff
@@ -149,11 +153,17 @@ public async Task<ActionResult> UpdateTeam([FromRoute] string aadGroupId)
                 { "IncomingAadGroupId", aadGroupId },
             };
 
+            // get the kronos org job mapping information
+            var teamMappings = await this.teamDepartmentMappingProvider.GetMappedTeamDetailsAsync(aadGroupId).ConfigureAwait(false);
+
+            // get the timezone for the first mapping - logically all mappings for the team must have the same time zone set
+            var timeZone = teamMappings[0].KronosTimeZone;
```
Comment (ajs117) [inline Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs]: what happens if teamMappings is null?

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs [hunk 5]
```diff
@@ -149,11 +153,17 @@ public async Task<ActionResult> UpdateTeam([FromRoute] string aadGroupId)
...
+
             // Check if payload is for open shift request.
             if (jsonModel.Requests.Any(x => x.Url.Contains("/openshiftrequests/", StringComparison.InvariantCulture)))
             {
                 // Process payload for open shift request.
-                responseModelList = await this.ProcessOpenShiftRequest(jsonModel, updateProps, aadGroupId, isRequestFromCorrectIntegration).ConfigureAwait(false);
+                responseModelList = await this.ProcessOpenShiftRequest(jsonModel, updateProps, aadGroupId, isRequestFromCorrectIntegration, timeZone).ConfigureAwait(false);
```
Comment (ajs117) [inline Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs]: Prefer kronosTimeZone variable name. Please check whole PR for where applicable.

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs [hunk 6]
```diff
@@ -716,8 +730,8 @@ private async Task<List<ShiftsIntegResponse>> ProcessSwapShiftRequestApprovalAsy
                 var newShiftSecond = JsonConvert.DeserializeObject<Shift>(postedShifts.Last().Body.ToString());
 
                 // Step 1 - Create the Kronos Unique ID.
-                var kronosUniqueIdFirst = this.utility.CreateUniqueId(newShiftFirst);
-                var kronosUniqueIdSecond = this.utility.CreateUniqueId(newShiftSecond);
+                var kronosUniqueIdFirst = this.utility.CreateUniqueId(newShiftFirst, timeZone);
+                var kronosUniqueIdSecond = this.utility.CreateUniqueId(newShiftSecond, timeZone);
```
Comment (ajs117) [inline Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs]: What happens if there are no postedShifts?

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TimeOffRequestsController.cs [hunk 7]
```diff
@@ -175,13 +168,17 @@ internal async Task ProcessTimeOffRequestsAsync(string isRequestFromLogicApp)
...
+                            if (timeOffRequestItems?.Count > 0)
                             {
+                                // get the team mappings for the team and pick the first because we need the Kronos Time Zone
+                                var teamMappings = await this.teamDepartmentMappingProvider.GetMappedTeamDetailsAsync(team).ConfigureAwait(false);
+                                var teamMapping = teamMappings.FirstOrDefault();
+
                                 foreach (var item in timeOffRequestItems)
                                 {
-                                    var timeOffReqStartDate = this.utility.UTCToKronosTimeZone(item.StartDateTime);
+                                    var timeOffReqStartDate = this.utility.UTCToKronosTimeZone(item.StartDateTime, teamMapping.KronosTimeZone);
```
Comment (ajs117) [inline Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TimeOffRequestsController.cs]: What happens if teamMapping is null?

**Review by ajs117 — CHANGES_REQUESTED** [review]

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/OpenShiftRequestController.cs (hunk 2 above)
Comment (Andy65) [reply Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/OpenShiftRequestController.cs]: The mappings between a team and a Kronos organisation job are fundamental to the working of the connector, so there should never be a time when no mappings are returned, however, it is a fair comment

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/SwapShiftController.cs (hunk 3 above)
Comment (Andy65) [reply Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/SwapShiftController.cs]: Modified to use the same approach as the OpenShiftRequestController to fall back to the global app setting for the time zone, however, as mentioned the mappings are fundamental to the working of the connector and if they are missing the connector will probably not work at all for the team.

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs (hunk 4 above)
Comment (Andy65) [reply Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs]: Again have switched it to the same as the other controllers with the same caveat regarding mappings.

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs (hunk 5 above)
Comment (Andy65) [reply Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs]: Fair enough

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs (hunk 6 above)
Comment (Andy65) [reply Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs]: Then it is not a valid shift swap from Teams - they are part of the contract - if a shift swap is approved then teams supplies a payload with 2 deleted shifts being the old shifts in the swap and 2 created shifts being the swapped shifts

File: Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TimeOffRequestsController.cs (hunk 7 above)
Comment (Andy65) [reply Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TimeOffRequestsController.cs]: Same change applied as elsewhere and same issue with missing mappings.

**Review by ajs117 — APPROVED** [review]

**Pull request merged**